crates/flux-middle/src/rty/expr.rs

```diff
@@ -674,13 +682,36 @@ impl Ctor {
     }
 }
 
+/// Given a primop `op` with signature `(t1,...,tn) -> t`
+/// We define a refined type for `op` expressed as a [`RuleMatcher`]
+///
+/// ```rust
+/// op :: (x1: t1, ..., xn: tn) -> { t[op_val[op](x1,...,xn)] | op_rel[x1,...,xn] }
+/// ```
+/// That is, using two *uninterpreted functions* `op_val` and `op_rel` that respectively denote
+/// 1. The _value_ of the primop, and
+/// 2. Some invariant _relation_ that holds for the primop.
+///
+/// The latter can be extended by the user via a `property` definition, which allows us
+/// to customize primops like `<<` with extra "facts" or lemmas. See `tests/tests/pos/surface/primops00.rs` for an example.
+///
+#[derive(Clone, PartialEq, Eq, Hash, TyEncodable, Debug, TyDecodable)]
+pub enum PrimFunc {
+    /// UIF representing the value of a primop
+    Val(BinOp),
+    /// UIF representing the relationship of a primop
+    Rel(BinOp),
+}
+
 #[derive(Clone, PartialEq, Eq, Hash, TyEncodable, Debug, TyDecodable)]
 pub enum ExprKind {
     Var(Var),
     Local(Local),
     Constant(Constant),
     ConstDefId(DefId),
     BinaryOp(BinOp, Expr, Expr),
+    /// An UIF application representing a primitive operation
+    PrimApp(PrimFunc, Expr, Expr),
```

Comment:
Or is it better to make `PrimFunc` a variant of `SpecFuncKind` ? I think so now...

Reply:
This is now split out. 

In `rty` we just use plain `App` everywhere _but_ have separate `GlobalFunc` (stuff that goes to fixpoint) and `InternalFunc` (stuff that does not); the latter also includes `Char/Int` etc. which are pulled out of `ThyFunc`.

In `fhir` we have a separate `PrimApp`, that uses the `BinOp` (to avoid shenanigans mangling the `BinOp` with the `Path` stuff needed in regular `App`).